### ISSUE TITLE
WarningsFinder fails with multiple similar methods

### ISSUE THREAD

**Issue by TWiStErRob** [mockito/mockito#581]
I found a bug in `org.mockito.internal.debugging.WarningsFinder#find` which is demonstrated by this minimal repro:
```java
interface Resources {
	String getString(int id);
	String getString(int id, Object... args);
}

public class MockitoBug {
	@Rule public MockitoRule mockito = MockitoJUnit.rule();
	@Mock Resources res;
	@Before public void stubResources() {
		when(res.getString(anyInt())).thenReturn("mock");
		when(res.getString(anyInt(), anyVararg())).thenReturn("mock");
	}
	@Test public void test() {
		res.getString(1, "forcedVarargOverload");
		res.getString(2, "forcedVarargOverload");
		fail("Needed so WarningsFinder.find is called."); // only brings it to the surface < 104-beta
	}
}
```
With the above code
`allInvocations` contains:
 * `res.getString(1, "forcedVarargOverload");`
 * `res.getString(2, "forcedVarargOverload");`

`unusedStubs` contains:
 * `res.getString(0);`

There are two nested `while` loops with iterators in `WarningsFinder`. The outer iteration can match multiple `hasSimilarMethod` so `Iterator.remove` can be called multiple times. So with the above data both `getString(1,...` and `getString(2,...` will match `getString(0)` and `unusedIterator.remove()` will be called twice.

If you run this with 103-beta or earlier you get the following exception:
```
java.lang.IllegalStateException
	at java.util.LinkedList$ListItr.remove(LinkedList.java:923) // Java 1.8.0b66 Win10 x64
	at org.mockito.internal.debugging.WarningsFinder.find(WarningsFinder.java:36)
	at org.mockito.internal.debugging.WarningsPrinterImpl.print(WarningsPrinterImpl.java:28)
	at org.mockito.internal.debugging.WarningsCollector.getWarnings(WarningsCollector.java:31)
	at org.mockito.internal.junit.JUnitRule$1.evaluate(JUnitRule.java:31) // logger.log(c.getWarnings());
```

This report may be irrelevant, because I couldn't find a way to reproduce this because #555 (fixing #384) removed the [line that calls `WarningsFinder`](https://github.com/mockito/mockito/commit/c3b4b9600f5434a67a3c083dc800763c7e7c5dbf#diff-6dc6ccd04e6127b2205ebddc98ee9b6dL31). In fact, it seems that `WarningsFinder` is not used any more in favor of the new classes added in that PR.

@szczepiq, I still suggest you take a look at it: I think this bug is possibly fixed by #384, but while the `WarningsFinder` class exists and not fixed, this bug is valid.

**Comment by szczepiq** [comment]
Wow, thanks a lot for the investigation!

Given that WarningsFinder is:
1. Only used by deprecated 'ConsoleSpammingRunner' and deprecated 'VerboseJUnitRunner' 
2. not part of public API

Do you think it is worth fixing?

**Comment by TWiStErRob** [comment]
1. + 2. A week ago that wasn't the case... :) Why do they still exist after #555? Were they just deprecated in 2.x? It crashed some of my real tests without accessing internals when using the rule. It will keep crashing them because I'm stuck at 42-beta because of PowerMock. So for me it's part of the public interface.

It's not a big change to the class, I think all is needed is to flag for removal in the inner loop and call `unusedIterator.remove()` right after the inner loop if flagged + a single test that covers it.

> Do you think it is worth fixing?

I guess it all depends on how long you're planning to keep this class. A week ago I would've said definitely, but now that you've rewritten the code, I'm not so sure. I guess it won't be affecting pretty much anyone in the final 2.0. If someone is using those runners, their tests must be working. It's probably quite an edge case with varargs, other than Android Resources and SLF4J Logger I can't think of any similar API. Both of these are commonly used, but likely not mocked often.

**Comment by bric3** [comment]
I agree with @szczepiq this does not worth spending time on this. I will close the issue as won't fix
